### PR TITLE
Add additional IG assertion checking

### PR DESCRIPTION
**Pull request by BruceForstall**
Especially, check various IG flags to ensure illegal combinations do not occur.

### CODE REVIEW

**Comment by BruceForstall** [comment]
I'd like to change the JIT so that an IG and all its EXTEND IGs (if any) have the same NOGC flag. I think it would simplify the mental model, and possibly reduce the chance of GC holes when peephole optimizations reach across IG boundaries. However, that requires further changes.

**Comment by BruceForstall** [comment]
[No diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=136343&view=ms.vss-build-web.run-extensions-tab), of course.

**Comment by BruceForstall** [comment]
@kunalspathak @dotnet/jit-contrib PTAL

File: src/coreclr/jit/emit.cpp [hunk 1]
```diff
@@ -1610,18 +1610,22 @@ void* emitter::emitAllocAnyInstr(size_t sz, emitAttr opsz)
...
-void emitter::emitCheckIGoffsets()
+// 2. Total code size
+// 3. IG flags
+//
+void emitter::emitCheckIGList()
 {
     size_t currentOffset = 0;
 
-    for (insGroup* tempIG = emitIGlist; tempIG != nullptr; tempIG = tempIG->igNext)
+    for (insGroup *tempIG = emitIGlist, *prevIG = nullptr; tempIG != nullptr; prevIG = tempIG, tempIG = tempIG->igNext)
```
Comment (kunalspathak) [inline src/coreclr/jit/emit.cpp:1628]: rename `tempIG` to `currIG`?

File: src/coreclr/jit/emit.cpp [hunk 2]
```diff
@@ -1630,12 +1634,82 @@ void emitter::emitCheckIGoffsets()
...
+        if (prevIG == nullptr)
+        {
+            // First IG can't be an extension group.
+            assert((tempIG->igFlags & IGF_EXTEND) == 0);
+
+            // First IG must be the function prolog.
+            assert(tempIG == emitPrologIG);
+        }
+
+        assert(emitPrologIG != nullptr);
```
Comment (kunalspathak) [inline src/coreclr/jit/emit.cpp:1647]: This doesn't have to be inside the loop.

**Review by kunalspathak — CHANGES_REQUESTED** [review]

**Comment by BruceForstall** [comment]
@kunalspathak updated

**Comment by BruceForstall** [comment]
diffs failure was infra:
```
git --config-env=http.extraheader=env_var_http.extraheader fetch --force --tags --prune --prune-tags --progress --no-recurse-submodules origin --depth=20 +3f93a90fc6171f75c303289c686b29b336f181bf:refs/remotes/origin/3f93a90fc6171f75c303289c686b29b336f181bf
##[error]The operation was canceled.
Unhandled exception. System.IO.IOException: The parameter is incorrect.
   at System.ConsolePal.ControlCHandlerRegistrar.Unregister()
   at System.Console.remove_CancelKeyPress(ConsoleCancelEventHandler value)
   at Agent.PluginHost.Program.Main(String[] args) in D:\a\1\s\src\Agent.PluginHost\Program.cs:line 202
```

**Review by kunalspathak — APPROVED** [review]

**Pull request merged**